soroban-sdk/src/crypto/bls12_381.rs

```diff
@@ -101,6 +158,61 @@ impl_bytesn_repr!(G2Affine, 192);
 impl_bytesn_repr!(Fp, 48);
 impl_bytesn_repr!(Fp2, 96);
 
+impl Fp {
+    pub fn env(&self) -> &Env {
+        self.0.env()
+    }
+
+    /// Maps to a `G1Affine` point via [simplified SWU
+    /// mapping](https://www.rfc-editor.org/rfc/rfc9380.html#name-simplified-swu-for-ab-0)
+    pub fn map_to_g1(&self) -> G1Affine {
+        self.env().crypto().bls12_381().map_fp_to_g1(self)
+    }
+}
+
+impl Into<BigInt<6>> for Fp {
+    fn into(self) -> BigInt<6> {
+        let inner: Bytes = self.0.into();
+        let mut limbs = [0u64; 6];
+        for i in 0..6u32 {
+            let start = i * 8;
+            let mut slice = [0u8; 8];
+            inner.slice(start..start + 8).copy_into_slice(&mut slice);
+            limbs[5 - i as usize] = u64::from_be_bytes(slice);
+        }
+        BigInt(limbs)
+    }
+}
+
+impl Neg for Fp {
+    type Output = Fp;
+
+    fn neg(self) -> Self::Output {
+        let fp_bigint: BigInt<6> = (&self.0).into();
+        if fp_bigint.is_zero() {
+            return self;
+        }
+
+        // BLS12-381 base field modulus
+        let mut res = BigInt([
+            13402431016077863595,
+            2210141511517208575,
+            7435674573564081700,
+            7239337960414712511,
+            5412103778470702295,
+            1873798617647539866,
+        ]);
+        // Compute modulus - value
+        let borrow = res.sub_with_borrow(&fp_bigint);
+        if borrow {
+            sdk_panic!("invalid input - Fp is larger than the field modulus")
+        }
```

Comment:
added a validity check. it's necessary because all the bls sdk types are just wrapper types of fixed-length bytes, with validity check performed on the host side. since here we are performing arithmetic directly on the bytes input (by-passing host) we need the check here. 
this check (on neg `Fp`) should be sufficient because all other type (`Fp2`, `G1`, `G2`) negations are boiled down to negation on `Fp`.

Reply:
that's a good suggestion, done